api/agent/call.go

```diff
@@ -406,13 +390,32 @@ func (c *call) Extensions() map[string]string {
 }
 
 func (c *call) RequestBody() io.ReadCloser {
-	if c.req.Body != nil && c.req.GetBody != nil {
-		rdr, err := c.req.GetBody()
-		if err == nil {
-			return rdr
-		}
+	// goal: we want to leave the write position of the cached body at the end,
+	// so that any request that manages to read bytes from the end of the original
+	// body gets tee'd onto the end of the cached body. meaning we always need to return
+	// a tee reader that tees all bytes read from the original body to the cached body,
+	// but reads only a copy of the cached body so that we don't have to move the read/write pointer,
+	// meaning we never call Read() on cachedBody
+	// TODO this could be optimized in the e.g payload case it's a strings reader, could check the type
+	// of the originalBody and futz with it if we really wanted, not v important
+	// TODO additionally, we only need to do this for lb_agent (we could use isLB bit, but maybe
+	// it would be better to leave this wart to encourage making the agent.Call interface a proper
+	// boundary and getting rid of that bit instead, so here!)
+
+	// TODO(reed): it's doubtful we need lock safety for this (but hey, here's a comment...)
+	if c.cachedBody == nil {
+		// XXX(reed): we could recycle these from End() which covers many of them
+		buf := bufPool.Get().(*bytes.Buffer)
+		buf.Reset()
+		c.cachedBody = buf
 	}
-	return c.req.Body
+
+	reader := io.TeeReader(c.originalBody, c.cachedBody)
+	// read anything in our cache first, then read the rest of the original body
+	// (which will get added to our cached body for the next time around)
+	reader = io.MultiReader(bytes.NewReader(c.cachedBody.Bytes()), reader)
+
+	return ioutil.NopCloser(reader)
```

Comment:
this may or may not be racy, which makes me think that it may be, it's kinda slick vs the previous tho so i'm trying to run with it. the properties are nice vs. the old read the entire thing into the buffer before sending (and extra goroutine to copy & timeout on), but I fear ) a data race on the underlying byte slice & ) a race where if 2 threads successfully read from the reader, say, 1 timed out but read to the teereader past the deadline, the second could end up with a 'gap' in the body (and this one in particular a mutex would not necessarily fix unless we serialized all readers in their entirety). thinking.

Reply:
I like this too, but I think it is indeed not safe. I'll take a look at it again when I'm rested, but maybe we should just accept the difficulty of this and just block and fill the buffer. We lose the nice streaming behavior and possibly take some hit on latency, but in practice I don't think this matters.

Reply:
i think i'm gonna make a reader that on the first read locks & dumps the entire req into a buffer (as you say), this is still a slight change in behavior from blocking & reading the entire request body into the buffer _before_ sending the body bytes (reduces memory overhead / latency in lb, esp in cases where placement fails entirely), as well as changing the abstraction to wrap things up into an io.Reader (this won't require an extra goroutine, since io is done elsewhere respecting the ctx timeout we won't have to explicitly do this). in any case, may be able to keep the teereader, but need to fiddle a bit.